apps/prairielearn/src/middlewares/authzWorkspace.js

```diff
@@ -16,11 +16,21 @@ const authzHasCoursePreviewOrInstanceView = promisify(
   require('./authzHasCoursePreviewOrInstanceView'),
 );
 
+const sql = sqldb.loadSqlEquiv(__filename);
+
 module.exports = asyncHandler(async (req, res, next) => {
   // We rely on having res.locals.workspace_id already set to the correct value here
-  const result = await sqldb.queryOneRowAsync(sql.select_auth_data_from_workspace, {
+  const result = await sqldb.queryZeroOrOneRowAsync(sql.select_auth_data_from_workspace, {
     workspace_id: res.locals.workspace_id,
   });
+
+  if (result.rows.length === 0) {
+    // We couldn't find the workspace. Someone could have put in a bad workspace ID,
+    // or there could be a dangling workspace after a variant was deleted. Either way,
+    // translate this to a 404 to keep the error out of our monitoring.
+    throw error.make(404, 'Workspace not found');
```

Comment:
In other middlewares, if the query being searched for is not found, we return a 403 instead. Any reason why here you use 404? Would 403 remain in the monitoring list?

Reply:
@mwest1066 remind me why we do that - is that so that someone can't check for the existence of resourced by differentiating between "this thing doesn't exist" and "I don't have access to it"?

Reply:
Neither 403s nor 404s are treated as errors by any of our monitoring, so from that perspective it doesn't matter much. I was going off of what GitHub does (renders a 404 if something isn't found or if you don't have permissions to view it); I either forgot or didn't realize that we had a different convention 🙃

Reply:
Yep, this was intended to avoid leaking information by having 403/404 values returned in different cases. I don't really care whether we always return 403 or always return 404, but we just don't want a mix of them.

Reply:
So I think we should return 403 here, just to be clear.

Reply:
As written, I think we'd have different error messages for "not found" vs "not authorized, which I think leaks as much information either way. I imagine we'd want to change this to just "Access denied" to match what all our authz middleware does, right?

Reply:
Yep, 'Access denied' is perfect to match our existing standard for this.

Reply:
Done in 1ade11b.